cou/apps/core.py

```diff
@@ -40,3 +45,117 @@ class Octavia(OpenStackApplication):
     """
 
     wait_timeout = 30 * 60  # 30 min
+
+
+@AppFactory.register_application(["nova-compute"])
+class NovaCompute(OpenStackApplication):
+    """Nova Compute application.
+
+    Nova Compute must wait for the entire model to be idle before declaring the upgrade complete.
+    """
+
+    wait_timeout = 30 * 60  # 30 min
+    wait_for_model = True
+    force: bool = False
+
+    def upgrade_steps(
+        self, target: OpenStackRelease, units: Optional[list[ApplicationUnit]]
+    ) -> list[UpgradeStep]:
+        """Upgrade steps planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param units: Units to generate upgrade steps,
+        :type units: Optional[list[ApplicationUnit]]
+        :raises ApplicationError: If no units are passed to upgrade Nova Compute
+        :return: List of upgrade steps.
+        :rtype: list[UpgradeStep]
+        """
+        if not units:
+            raise ApplicationError("No units passed to upgrade Nova Compute.")
+        app_steps = super().upgrade_steps(target, units)
+        unit_steps = self._get_units_upgrade_steps(units)
+        return app_steps + unit_steps
+
+    def _get_units_upgrade_steps(self, units: list[ApplicationUnit]) -> list[UpgradeStep]:
+        """Get the upgrade steps for the units.
+
+        :param units: Units to generate upgrade steps
+        :type units: list[ApplicationUnit]
+        :return: List of upgrade steps
+        :rtype: list[UpgradeStep]
+        """
+        units_steps = UpgradeStep(
+            description=f"Upgrade plan for units: {', '.join([unit.name for unit in units])}",
+            parallel=True,
+        )
+
+        for unit in units:
+            unit_steps = UnitUpgradeStep(description=f"Upgrade plan for unit: {unit.name}")
+            unit_steps.add_step(self._get_disable_scheduler_step(unit))
+            unit_steps.add_step(self._get_empty_hypervisor_step(unit))
+            unit_steps.add_step(self._get_pause_unit_step(unit, self._dependency_of_steps))
+            unit_steps.add_step(self._get_openstack_upgrade_step(unit, self._dependency_of_steps))
+            unit_steps.add_step(self._get_resume_unit_step(unit, self._dependency_of_steps))
```

Comment:
```suggestion
            unit_steps.add_step(self._get_pause_unit_step(unit, not self.force))
            unit_steps.add_step(self._get_openstack_upgrade_step(unit,  not self.force))
            unit_steps.add_step(self._get_resume_unit_step(unit,  not self.force))
```
And even for force I start to have feeling, that it should be argument for `upgrade_steps` and not class argument.

Reply:
The problem is that `force` is very specific for `NovaCompute` and the `upgrade_steps` is inherited from `OpenStackApplication`. Adding a new parameter for just one class does not looks right to me.

IMO `not self.force` is very out of context. Having a property adds a little overhead, but it's more expressive.